src/transport/Tuple.h

```diff
@@ -0,0 +1,215 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *
+ * Defines a compound transport type (tuple) that can merge several transports
+ * to look like a single one.
+ */
+#ifndef TRANSPORT_TUPLE_H_
+#define TRANSPORT_TUPLE_H_
+
+#include <tuple>
+#include <type_traits>
+
+#include <transport/Base.h>
+
+namespace chip {
+namespace Transport {
+
+/**
+ * Groups together several transports of different types and presents them as a unified transport.
+ *
+ * The usage intent is to be able to group several distinct transport types and make them look
+ * as a single transport.
+ *
+ * Having an example class definition of:
+ *    Tuple<UDP, UDP, TCP>
+ *
+ * Is equivalent to:
+ *    class TupleOfUdpUdpTcp : public BASE {
+ *       private:
+ *          UDP mUdp1;
+ *          UDP mUdp2;
+ *          TCP mTcp2;
+ *       public:
+ *          Init(UDPListenParameters &, UDPListenParameters&, TCPListenParameters) {...}
+ *          CHIP_ERROR SendMessage(...) override;
+ *          bool CanSendToPeer(...) override;
+ *    }
+ *
+ * The intent of this is to allow applications to use any transport types without CHIP pre-defining
+ * popular mappings (like UDP only, UDP and BLE, BLE only etc.) and without using #ifdefs to create
+ * a single 'standard transport'.
+ *
+ * Transport logic:
+ *   - Every transport can decide if a 'PeerAddress' can be sent over 'self'
+ *
+ *   - Within a mixed tuple, if several transports support a peer address (e.g. TCP and UDP both
+ *     support IP), the first one found wins.
+ *
+ *   - Expected message sending logic:
+ *     - BLE transports only support BLE. TCP/UDP support IP addresses
+ *     - Multicasts only supported by UDP
+ *
+ * @param TransporTypes the ORDERED list of transport types grouped together. Order matters in
+ *        determining what transport is used when multiple transports can reach a Peer.
+ *
+ * Transports (TransportTypes) are assumed to have an Init call with EXACTLY one argument and returning
+ * a CHIP_ERROR, with a signature like:
+ *
+ *    CHIP_ERROR Init(AnyType);
+ *
+ */
+template <typename... TransportTypes>
+class Tuple : public Base
+{
+public:
+    /** @overload */
```

Comment:
Thanks. When running `make doc`, does the right thing happen or does Doxygen squawk?

Reply:
make doc yells at me as 'does not exist' and 'docdist' also complains. There is a makefile under a 'docs' directory however make in there does not do anything.

I am not sure how to build docs :( Still looking though.

Reply:
created #1486 for this: just tried on master checkout and help claims `make doc` builds docs, however instead I get an error.

Reply:
Got the docs to build and as a result I made some code formatting changes.

Also found that @overload does more harm than good: if I add it, doxygen adds automatically a 'This is an overloaded member function, provided for convenience. It differs from the above function only in what argument(s) it accepts.' where as if I do not add it, it copies over the doxygen from the base class (which seems to be what we want.

I dropped overloads as a result.

Reply:
Great, thanks for checking it out.